### PR TITLE
Reorganize Cockroach startup to eliminate ExecStartPost timeouts

### PR DESCRIPTION
**Pull request by jongiddy**
## High-level description

Remove the ExecStartPost call to cockroach-change-config script in the Cockroach service and replace with a dependency to the config change unit. This fixes a problem where CockroachDB accepts the connection from the script but never gets to a state ready to receive data, so the unit times out. To make the script run at the right time, we use systemd dependencies, and reduce the chance of units failing, so that they run in the desired order.

To do this, we create a chain of dependencies, Exhibitor (ZooKeeper) -> CA (Enterprise only) -> Cockroach -> Bouncer -> Admin Router. Each of these needs the previous service to be available. This reduces the number of failure loops during installation and reboot.

I've removed the "systemd rules" doc file, since we're definitely breaking those rules here. The file is over 4 years old and there's no reference to known bugs or other rationale for the rules. If there were bugs, they are likely to be fixed in deployed platforms now. If not, we need to rediscover them so we can justify these constraints.

## Corresponding DC/OS tickets (required)

  - [D2IQ-62292](https://jira.d2iq.com/browse/D2IQ-62292) Remove possibility of timeouts in systemd ExecStartPost
  - [D2IQ-51276](https://jira.d2iq.com/browse/D2IQ-51276) Consoldiate CA bootstrap? (CA bootstrap does not retry or fail on Docker reload failure)

## Related tickets (optional)

- https://jira.d2iq.com/browse/COPS-5601
- https://jira.d2iq.com/browse/COPS-5795
- https://jira.d2iq.com/browse/COPS-6063

### CODE REVIEW

**Comment by jongiddy** [comment]
@mesosphere-mergebot test all

**Comment by jongiddy** [comment]
@mesosphere-mergebot test teamcity/dcos/build/dcos

**Comment by jongiddy** [comment]
@mesosphere-mergebot test teamcity/dcos/test/terraform/aws/onprem/static/group1

**Comment by jongiddy** [comment]
@mesosphere-mergebot test all

**Comment by jongiddy** [comment]
@mesosphere-mergebot  test teamcity/dcos/build/dcos

**Comment by jongiddy** [comment]
@mesosphere-mergebot test teamcity/dcos/build/dcos

**Comment by jongiddy** [comment]
@mesosphere-mergebot test all

**Comment by jongiddy** [comment]
@mesosphere-mergebot test all

**Comment by jongiddy** [comment]
@mesosphere-mergebot request-review @jkoelker @janisz @mhrabovcin

**Comment by janisz** [comment]
I'm not sure if this should happen in this PR but how about merging EE and OSS Cocroach package how we did for marathon and metronome?

File: packages/cockroach/extra/dcos-cockroach.service [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
+Documentation=https://docs.d2iq.com/mesosphere/dcos/
+# Wait for ZooKeeper to be available
+After=dcos-exhibitor.service
+# When CockroachDB is started, run the config change immediately afterwards
+Wants=dcos-cockroachdb-config-change.service
+Before=dcos-cockroachdb-config-change.service
 
 [Service]
-Type=simple
+Type=notify
```
Comment (jkoelker) [inline packages/cockroach/extra/dcos-cockroach.service:11]: TIL

**Review by jkoelker — DISMISSED** [review]
and there was much rejoicing
and there was much rejoicing

**Comment by jongiddy** [comment]
@mesosphere-mergebot label Work In Progress

**Comment by jongiddy** [comment]
@mesosphere-mergebot label Ready For Review

**Review by jkoelker — APPROVED** [review]

File: packages/exhibitor/extra/dcos-exhibitor.service [hunk 2]
```diff
@@ -35,3 +36,5 @@ ExecStartPre=$PKG_PATH/bin/bootstrap_exhibitor_tls.py
 ExecStartPre=$PKG_PATH/bin/set_exhibitor_file_permissions.py
 # Start Exhibitor
 ExecStart=$PKG_PATH/usr/exhibitor/start_exhibitor.py
+# Wait for ZooKeeper to start listening
+ExecStartPost=-timeout 20 /bin/sh -c 'until lsof -i :2181; do sleep 1; done'
```
Comment (jkoelker) [inline packages/exhibitor/extra/dcos-exhibitor.service:40]: Looks like the test wants this to be `/usr/bin/timeout`

Comment (jkoelker) [reply packages/exhibitor/extra/dcos-exhibitor.service]: I went ahead and changed the path and pushed to the branch will approve again once the tests pass again.

**Review by jkoelker — APPROVED** [review]

**Comment by jongiddy** [comment]
Missing notification from teamcity/dcos/test/terraform/aws/onprem/static/group3 and teamcity/dcos/test/test-e2e/group1

**Comment by jongiddy** [comment]
@mesosphere-mergebot label Ship It

**Pull request merged**